### DISCUSSION TITLE
OutOfMemoryError on BM25 Document Retrieval

### DISCUSSION THREAD

**Discussion by limelody** [castorini/pyserini#1299] [Q&A]
I’m getting a Java OOM error on the indexing stage of [this](https://github.com/castorini/pyserini/blob/master/docs/experiments-msmarco-doc.md) experiment, my system configs are quite close though (8 GB RAM, 140GB of SSD left).

I tried adjusting the ‘—memorybuffer’ parameter too but my system always fails at ~500k documents. Is there any other parameter I could adjust to let the indexing complete, or is it likely just because my SSD is cutting it close?

**Comment by GioiaZheng** [comment]
This is JVM heap, not disk — 140 GB free SSD isn't the issue, and the crash at ~500k docs is the giveaway.

Two things to fix:

1. **Don't raise `--memorybuffer`, lower it.** It's the Lucene `IndexWriter` RAM buffer (MB) and it's allocated *inside* the heap, so a large value makes the OOM happen sooner. Try `--memorybuffer 2048` or less.

2. **Raise the JVM max heap.** Pyserini boots the JVM on first import, so you have to set heap *before* that happens:

```python
import jnius_config
jnius_config.add_options('-Xmx6g')
import pyserini  # JVM starts here — options after this are ignored
```

On 8 GB physical RAM keep `-Xmx` around 6g and leave the rest for the OS / page cache.

Also drop `--threads` if you raised it — each indexing thread holds its own buffer, so threads × memorybuffer is what actually blows the heap.
